### PR TITLE
FLEURS ASR Recipe for ESPnet2

### PR DESCRIPTION
**Pull request by wanchichen**
This PR is for the creation of an ASR recipe for the FLEURS dataset, a speech version of the FLORES101 MT dataset.

The dataset can only be downloaded from the [FLEURS Hugging Face Dataset](https://huggingface.co/datasets/google/xtreme_s#1-speech-recognition-asr), which requires the installation of the datasets library (commit 8757bf666bf34507851cf90752c0e97f3694e92e).

This PR supports monolingual ASR and 102-way multilingual ASR. This thus requires a large amount of language id tokens needed for multilingual ASR. A [change](https://github.com/wanchichen/espnet/blob/master/egs2/TEMPLATE/asr1/asr.sh#L645-L651) is proposed to the ASR template so that the bpe_nlsyms flag can also support a file of language id tokens, in addition to a comma-delimited list.

### CODE REVIEW

File: egs2/fleurs/asr1/local/nlsyms.txt [hunk 1]
```diff
@@ -0,0 +1,103 @@
+af_za
```
Comment (ftshijt) [inline egs2/fleurs/asr1/local/nlsyms.txt:1]: Is it possible to generate automatically? It is no problem to host here, but if possible generating automatically, could be better. You may refer this for creating the id

**Review by ftshijt — CHANGES_REQUESTED** [review]
Thanks for your contribution! Some minor issues or suggestions as follows:

File: egs2/fleurs/asr1/README.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+|dataset|Snt|Wrd|Corr|Sub|Del|Ins|Err|S.Err|
+|---|---|---|---|---|---|---|---|---|
+|decode_asr_lm_lm_train_lm_en_us_bpe150_valid.loss.ave_asr_model_valid.acc.ave/dev_en_us|388|7935|5.8|84.3|10.0|7.6|101.8|100.0|
+|decode_asr_lm_lm_train_lm_en_us_bpe150_valid.loss.ave_asr_model_valid.acc.ave/test_en_us|647|14344|6.0|83.7|10.2|7.6|101.6|100.0|
+
+### CER
+
+|dataset|Snt|Wrd|Corr|Sub|Del|Ins|Err|S.Err|
+|---|---|---|---|---|---|---|---|---|
+|decode_asr_lm_lm_train_lm_en_us_bpe150_valid.loss.ave_asr_model_valid.acc.ave/dev_en_us|388|47051|41.3|26.5|32.3|6.6|65.3|100.0|
```
[the remark this answers is not in the document]
Comment (wanchichen) [reply egs2/fleurs/asr1/README.md:23]: Sorry forgot to mention, the results are very preliminary. Have not really explored ASR training and model parameters. There is about 75 hours of English data in the corpus.

File: egs2/fleurs/asr1/local/nlsyms.txt (hunk 1 above)
Comment (wanchichen) [reply egs2/fleurs/asr1/local/nlsyms.txt:1]: Yes, this is generated automatically as well when the dataset is downloaded in https://github.com/wanchichen/espnet/blob/master/egs2/fleurs/asr1/local/create_dataset.py#L108-L112

File: egs/commonvoice/asr1/local/data_prep.pl [hunk 3]
```diff
@@ -54,7 +54,7 @@
   }
   print TEXT "$uttId"," ","$text","\n";
   print GNDR "$uttId"," ","$gender","\n";
-  print WAV "$uttId"," ffmpeg -i $db_base/clips/$filepath -f wav -ar 16000 -ab 16 -ac 1 - |\n";
+  print WAV "$uttId"," ffmpeg -i $db_base/$filepath -f wav -ar 16000 -ab 16 -ac 1 - |\n";
```
[the remark this answers is not in the document]
Comment (wanchichen) [reply egs/commonvoice/asr1/local/data_prep.pl]: Thank you for the catching that. Was a mistake, fixed in e5fcb65bd95b07c5cffcf133e6801c9dd3b5fcaf

File: egs2/fleurs/asr1/local/nlsyms.txt (hunk 1 above)
Comment (ftshijt) [reply egs2/fleurs/asr1/local/nlsyms.txt:1]: Sounds cool! Then we do not have to include that in the PR?

File: egs2/fleurs/asr1/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (ftshijt) [reply egs2/fleurs/asr1/README.md:23]: I see, if it has 75 hours, the results seems not very satisfactory. But it indeed may happen with different data conditions and domains. My current suggestion is that 150 bpe seems to be small for an English model. You might want to increase it to have a try (though it could not be the fundamental issue for now.

File: egs2/fleurs/asr1/local/nlsyms.txt (hunk 1 above)
Comment (wanchichen) [reply egs2/fleurs/asr1/local/nlsyms.txt:1]: Yeah the code in the PR already supports it

File: egs2/fleurs/asr1/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (wanchichen) [reply egs2/fleurs/asr1/README.md:23]: Ok, I will take a look into the bpe as well. Will make update when training is finished

File: egs2/fleurs/asr1/local/create_dataset.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
+import argparse
+import traceback
+
+import pandas as pd
+
+try:
+    from datasets import load_dataset
```
Comment (sw005320) [inline egs2/fleurs/asr1/local/create_dataset.py:7]: This is a very good way to handle the import error.
In addition, you may also add this error handling to `local/path.sh` (See https://github.com/espnet/espnet/blob/master/egs2/how2_2000h/asr1/local/path.sh)

Comment (sw005320) [reply egs2/fleurs/asr1/local/create_dataset.py:7]: This is our standard way to handle the installation issue

File: egs2/fleurs/asr1/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (sw005320) [reply egs2/fleurs/asr1/README.md:23]: Can you paste the learning curve?
I think this training is just corrupted and we need to tune the optimization hyperparameters (just learning rate).

**Review by sw005320 — APPROVED** [review]
LGTM

**Comment by wanchichen** [comment]
loss
acc

These are the accuracy/loss plots. This is after I increased the bpe vocab size as recommended by @ftshijt .
I am re-training the model with a lowered learning rate.

**Comment by sw005320** [comment]
I think there are some critical issues not related to BPE.
Completely wrong optimization parameters or data preparation is wrong.
Can you check it with @brianyan918?

But the PR itself is good.
After you deal with some comments (`local/path.sh` and `nlsyms.txt`), I'll merge this PR.

**Comment by sw005320** [comment]
Which configs are you using?
`lr` seems to be high.

**Comment by wanchichen** [comment]
I copied over the asr_train_transformer config from common voice. The LR was 1.0. I'm retraining it with 0.0001.

**Comment by sw005320** [comment]
OK.
The WSJ recipe is similar (English, 80 hours), and you may refer it 
https://github.com/espnet/espnet/blob/master/egs2/wsj/asr1/conf/tuning/train_asr_transformer2.yaml
0.0001 might be too small.

Also, I'm concerned about some issues in the data preparation.
Please check whether the text data are well processed and the corresponding audio data is correct or not (by listening to a few samples).

**Comment by wanchichen** [comment]
I have addressed `local/path.sh` in the above commits. Do you want me to remove the `nlsyms.txt` file? It is auto-generated anyways when the user runs the download script.

I used the exact same configuration as https://github.com/espnet/espnet/tree/master/egs2/wsj/asr1#using-transformer-lm-asr-model-is-same-as-the-above-lm_weight12-ctc_weight03-beam_size20 for lm, asr, decoding, and inference. The results seem better, but still pretty bad. The results below are from averaging the 10 best models.

I listened to the audio for about 20 samples, both before and after any processing, as well as checked the text that its supposed to match to. No issues there either.


# RESULTS
## Environments

- date: `Mon Jun 20 11:24:32 EDT 2022`
- python version: `3.8.6 (default, Dec 17 2020, 16:57:01)  [GCC 10.2.0]`
- espnet version: `espnet 202205`
- pytorch version: `pytorch 1.8.1+cu102`
- Git hash: `45e8cb9173a072f85ee7a7ccbcae06af7c5c484a`
  - Commit date: `Wed Jun 1 14:21:14 2022 +0900`

## asr_train_asr_transformer_raw_en_us_bpe300_sp
### WER

|dataset|Snt|Wrd|Corr|Sub|Del|Ins|Err|S.Err|
|---|---|---|---|---|---|---|---|---|
|decode_transformer_lm_lm_train_lm_en_us_bpe300_valid.loss.ave_asr_model_valid.acc.ave/dev_en_us|388|7935|10.8|48.5|40.6|1.3|90.4|100.0|
|decode_transformer_lm_lm_train_lm_en_us_bpe300_valid.loss.ave_asr_model_valid.acc.ave/test_en_us|647|14344|11.3|45.9|42.8|1.2|89.9|100.0|

### CER

|dataset|Snt|Wrd|Corr|Sub|Del|Ins|Err|S.Err|
|---|---|---|---|---|---|---|---|---|
|decode_transformer_lm_lm_train_lm_en_us_bpe300_valid.loss.ave_asr_model_valid.acc.ave/dev_en_us|388|47051|34.1|22.4|43.5|4.0|69.8|100.0|
|decode_transformer_lm_lm_train_lm_en_us_bpe300_valid.loss.ave_asr_model_valid.acc.ave/test_en_us|647|83954|33.8|22.2|44.0|3.9|70.1|100.0|

### TER

|dataset|Snt|Wrd|Corr|Sub|Del|Ins|Err|S.Err|
|---|---|---|---|---|---|---|---|---|
|decode_transformer_lm_lm_train_lm_en_us_bpe300_valid.loss.ave_asr_model_valid.acc.ave/dev_en_us|388|22491|16.8|51.1|32.1|3.4|86.6|100.0|
|decode_transformer_lm_lm_train_lm_en_us_bpe300_valid.loss.ave_asr_model_valid.acc.ave/test_en_us|647|39965|16.5|51.2|32.3|3.4|86.9|100.0|

acc

loss
